src/shared/calendarspec.c

```diff
@@ -1251,11 +1276,13 @@ static int find_next(const CalendarSpec *spec, struct tm *tm, usec_t *usec) {
 
         c = *tm;
         tm_usec = *usec;
+        dst = spec->dst;
 
         for (unsigned iteration = 0; iteration < MAX_CALENDAR_ITERATIONS; iteration++) {
                 /* Normalize the current date */
                 (void) mktime_or_timegm_usec(&c, spec->utc, /* ret= */ NULL);
-                c.tm_isdst = spec->dst;
+                if (dst != -2)
```

Comment:
as requested before: just turn this -2 thing into a proper bool. i.e. `bool invalidate_dst; = false` or so.